Automattic/jetpack
# Add/jitm builder

# Justin Time Messages

~Requires:~ D5613-code ~to be available first~ shipped.

Uses an API to determine which JITM to show. At first it may seem like a bad idea to have an ajax call fire on every wp-admin page view. However, the call is extremely lightweight if there is not JITM defined for a page, bailing out nearly instantly.

It's designed to cache a call for 5 minutes by default (unless the JITM being shown sets something differently). If a sync occurs (`last_heartbeat` option, also adds a `last_sync` option) it will disregard the cache. This is useful in the case where a user clicks a CTA on a JITM or causes some action that would hide the JITM. Such as enabling Akismet. This seems to trigger a heartbeat instantly and we invalidate the cache. I'm open to any and all suggestions about making this more reliable and resilient.

- [x] ~Currently waiting on #7117~

#### Testing instructions:

- ~You'll need D5613-code, see that testing plan as well.~
- Make sure you do a `yarn build`
- Verify defined JITM's work as defined (matches `master` behavior)
- Ensure JITM's dismiss properly by `product_class` (if one VaultPress is dismissed, all of them do)
- Ensure activating Akismet will prevent the JITM from showing on the "Comments" screen.

#### Existing JITMs:

- Akismet: comments page with akismet deactivated
- VaultPress: update-core page and just after publishing a post with VaultPress deactivated
- WooCommerce: Woo installed and activated set to USA or Canada, on woo settings page (among others) ... woo services plugin must be deactivated or not installed.


#### Proposed changelog entry for your changes:
N/A
